builder/ecs/step_attach_volume.go

```diff
@@ -0,0 +1,289 @@
+		err = fmt.Errorf("error initializing EVS client: %s", err)
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+	imsClient, err := config.HcImsClient(region)
+	if err != nil {
+		err = fmt.Errorf("error initializing IMS client: %s", err)
+		state.Put("error", err)
+		return multistep.ActionHalt
+	}
+
+	serverId := state.Get("server_id").(string)
+	attachVolumeIds := make([]string, 0)
+	for _, dataVolumeWrap := range dataVolumeWraps {
+		switch dataVolumeWrap.dataType {
+		case VolumeId:
+			err = attachVolume(ui, state, ecsClient, serverId, dataVolumeWrap.DataVolume)
+			if err != nil {
+				state.Put("error", err)
+				return multistep.ActionHalt
+			}
+			attachVolumeIds = append(attachVolumeIds, dataVolumeWrap.VolumeId)
+		case Size:
+			err = createVolume(ui, state, evsClient, serverId, dataVolumeWrap.DataVolume)
+			if err != nil {
+				state.Put("error", err)
+				return multistep.ActionHalt
+			}
+		case DataImageId:
+			err = createVolumeByImageId(ui, state, imsClient, evsClient, serverId, dataVolumeWrap.DataVolume)
+			if err != nil {
+				state.Put("error", err)
+				return multistep.ActionHalt
+			}
+		case SnapshotId:
+			err = createVolumeBySnapshotId(ui, state, evsClient, serverId, dataVolumeWrap.DataVolume)
+			if err != nil {
+				state.Put("error", err)
+				return multistep.ActionHalt
+			}
+		}
+		state.Put("attach_volume_ids", attachVolumeIds)
+	}
+	return multistep.ActionContinue
+}
+
+func (s *StepAttachVolume) Cleanup(state multistep.StateBag) {
+}
+
+func parseDataVolumes(dataVolumes []DataVolume) ([]DataVolumeWrap, error) {
+	dataVolumeWraps := make([]DataVolumeWrap, 0, len(dataVolumes))
+	for _, dataVolume := range dataVolumes {
+		specified := make([]string, 0)
+		if dataVolume.Type == "" {
+			dataVolume.Type = "SSD"
+		}
+		if dataVolume.VolumeId != "" {
+			specified = append(specified, "volume_id")
+			dataVolumeWraps = append(dataVolumeWraps, DataVolumeWrap{dataVolume, VolumeId})
+		}
+		if dataVolume.Size > 0 {
+			specified = append(specified, "volume_size")
+			dataVolumeWraps = append(dataVolumeWraps, DataVolumeWrap{dataVolume, Size})
+		}
+		if dataVolume.DataImageId != "" {
+			specified = append(specified, "data_image_id")
+			dataVolumeWraps = append(dataVolumeWraps, DataVolumeWrap{dataVolume, DataImageId})
+		}
+		if dataVolume.SnapshotId != "" {
+			specified = append(specified, "snapshot_id")
+			dataVolumeWraps = append(dataVolumeWraps, DataVolumeWrap{dataVolume, SnapshotId})
+		}
+		if len(specified) == 0 {
+			return nil, fmt.Errorf("one of volume_id, volume_size, data_image_id, snapshot_id must be specified")
+		}
+		if len(specified) > 1 {
+			return nil, fmt.Errorf("only one of volume_id, volume_size, data_image_id, snapshot_id can be"+
+				"specified, but `%s` were specified", strings.Join(specified, ","))
+		}
+	}
+	return dataVolumeWraps, nil
+}
+
+func attachVolume(ui packer.Ui, state multistep.StateBag, ecsClient *ecs.EcsClient, serverId string, dataVolume DataVolume) error {
```

Comment:
it's better to define the function as `attachDataVolumes`

Reply:
done